feat: added select all button to customize columns modal; added styles; revised instruction text

kai-liu opened the pull request:
Contributes to
[https://github.com/carbon-design-system/ibm-cloud-cognitive/issues/1613](https://github.com/carbon-design-system/ibm-cloud-cognitive/pull/url)
[https://github.com/carbon-design-system/ibm-cloud-cognitive/issues/1981](https://github.com/carbon-design-system/ibm-cloud-cognitive/pull/url)

Implemented select all functionalities to the columns.
Implemented style changes to match the style of a Datagrid table.
Revised text to improve clarity.

What did you change?
within the Datagrid directory:

`DraggableElement.js` Added selected attribute to help with css selectors
`addons/CustomizeColumns/Actions.js` Removed search button, as required by design.

`CustomizeColumns/Columns.js` Added select-all element and implement logic to support it

`CustomizeColumns/CustomizeColumnsModal.js` Added logic to show number of visible columns vs total number of columns. Removed elements in accordance to the carbon design document [https://pages.github.ibm.com/cdai-design/pal/components/data-table/column-customization/usage/](https://github.com/carbon-design-system/ibm-cloud-cognitive/pull/url)

`/styles/addons/CustomizeColumnsModal.scss` Style changes.

`/styles/draggableElement.scss' Style changes

How did you test and verify your work?
storybook

matthewgallo commented on packages/cloud-cognitive/src/components/Datagrid/styles/addons/_CustomizeColumnsModal.scss:
```diff
@@ -13,30 +13,45 @@
 }
 
-.#{$block-class}__customize-columns-checkbox {
+.bx--form-item.bx--checkbox-wrapper.#{$block-class}__customize-columns-checkbox-wrapper {
```
Can you add the carbon prefix variable?
```scss
.#{$block-class} .#{$carbon-prefix}--form-item.#{$carbon-prefix}--checkbox-wrapper.#{$block-class}__customize-columns-checkbox-wrapper
```

kai-liu replied:
Removed the `form-item` and `checkbox-wrapper` selectors as they are unecessary

matthewgallo commented on packages/cloud-cognitive/src/components/Datagrid/styles/addons/_CustomizeColumnsModal.scss:
```diff
@@ -13,30 +13,45 @@
+  border-bottom: none;
+}
+div.#{$block-class}__customize-columns-modal--actions,
+div.bx--search-magnifier,
+.bx--search-magnifier-icon {
```
These also need to use the `$carbon-prefix` variable as well

kai-liu replied:
added `$carbon-prefix`

matthewgallo commented on packages/cloud-cognitive/src/components/Datagrid/styles/addons/_CustomizeColumnsModal.scss:
```diff
@@ -46,3 +61,22 @@
+  padding-left: $spacing-08;
+}
+
+.bx--checkbox-label {
```
Use `$carbon-prefix` and also this rule should be scoped to the datagrid component, ie
```scss
.#{$block-class} .#{$carbon-prefix}--checkbox-label {
...
}
```

kai-liu replied:
replaced `.bx--checkbox-label` with `.#{$block-class} .#{$carbon-prefix}--checkbox-label`

kai-liu replied:
Also moved the rule to `datagrid.scss`

matthewgallo commented on packages/cloud-cognitive/src/components/Datagrid/styles/addons/_CustomizeColumnsModal.scss:
```diff
@@ -46,3 +61,22 @@
+  }
+}
+
+#bx--modal-body--modal-2 {
```
This id changes each time the modal is rendered so it's not a reliable selector here. You could potentially use something like:
```scss
.#{$block-class} .#{$block-class}__customize-columns-modal .#{$carbon-prefix}--modal-content {
...
}
```

kai-liu replied:
replaced `#bx--modal-body--modal-2` with `.#{$block-class} .#{$block-class}__customize-columns-modal .#{$carbon-prefix}--modal-content`

kai-liu commented:
Comments from  https://github.com/carbon-design-system/ibm-cloud-cognitive/pull/2062 :

The convention is to use single quotes for string literals and package names, double quotes for React props or element attributes. Are the changes made by prettifier? Can you revert to single quotes (follow convention of existing code) and then do npm run lint to verify?

Seems like I was previously using the a different lint config that has double quotes by default. I have changed the lint config and used npm run lint to verify.

Why are we taking out "reset to default"? I think the request is to add a select all / unselect all checkbox in the customize column modal. It doesn't seem to mention that we need to remove the "reset to default" feature. Is this an explicit request from design?

The changes are made to meet the requirements as illustrated in the [design document ](https://pages.github.ibm.com/cdai-design/pal/components/data-table/column-customization/usage/). It was not an explicit request. I have discussed this with the PAL team, and was advised to follow the document as much as possible.

Can we further simplify the select / unselect all logic? On a high level, when the "select/unselect all" checkbox is clicked, there should be a for-loop that selects / unselects each column checkbox by calling the corresponding on change callback. The existing on change callback doesn't need to be changed. Also, we can probably implement a function which walks the column checkbox states and return a value that indicates if all/some/none column checkboxes are checked. This can be used to populate the select/unselect all checkbox state.

I have simplified the select/ unselect all logic, as well as reverted the change callback to the original callback. The logic uses a for-loop to check/uncheck each checkboxes. It makes the decision by traversing all checkboxes and collect their states. If all checkboxes are 'checked', then for loop will uncheck them, and vice versa. If some checkboxes are checked but not all, then the for loop will check them.

holmansze commented on packages/cloud-cognitive/src/components/Datagrid/Datagrid/addons/CustomizeColumns/CustomizeColumnsModal.js:
```diff
@@ -76,12 +80,29 @@ const CustomizeColumnsModal = ({
+    return result;
+  }, [columnObjects]);
+
   const string = searchText.trim().toLowerCase();
```
You may simplify this by something like:

```
columnObjects.filter((col) => col.isVisible).length
```

kai-liu replied:
Thanks for the suggestion. Replaced the code block with `columnObjects.filter((col) => col.isVisible).length`

kai-liu commented on packages/cloud-cognitive/src/components/Datagrid/Datagrid/addons/CustomizeColumns/Columns.js:
```diff
@@ -90,6 +136,25 @@ const Columns = ({
           </span>
+          <div
+            id={'customize-columns-select-all'}
+            className={setSelectAllColumnClassName()}
```
Replaced `className={setSelectAllColumnClassName()}` with `<div className={
   classNames({
      [`${blockClass}__draggable-handleHolder`]: getVisibleColumnsCount() === 0,
      [`${blockClass}__draggable-handleHolder-selected`]: getVisibleColumnsCount() > 0,
   })`

kai-liu commented on packages/cloud-cognitive/src/components/Datagrid/Datagrid/addons/CustomizeColumns/Columns.js:
```diff
@@ -54,6 +56,50 @@ const Columns = ({
+      }
+      li.setAttribute('selected', isVisible);
+    });
+  };
```
Removed `setColumnStatus` and `initializeList`. Revised logic to handle attribute values by passing a `selected` prop to the `DraggableElement` component during render. This reduces code clunkiness a lot. Thanks for the advice!

holmansze commented on packages/cloud-cognitive/src/components/Datagrid/Datagrid/addons/CustomizeColumns/Columns.js:
```diff
@@ -54,6 +57,19 @@ const Columns = ({
+      onSelectColumn(columns[i - 1], selectAll);
+    }
+  };
+
```
Instead of setting the `checked` attributes explicitly, update the `columns` array. Once the `columns` array is updated, React will re-render the checkboxes with the updated state.

The following code from `CustomizeColumnsModal.js` may be a useful reference. Take a look and see a similar technique can be applied here:
https://github.com/carbon-design-system/ibm-cloud-cognitive/blob/94f3ba91fefcc8acdcb8dc12d3db6d2287e0bc81/packages/cloud-cognitive/src/components/Datagrid/Datagrid/addons/CustomizeColumns/CustomizeColumnsModal.js#L65

kai-liu replied:
Removed `checkboxes[i].checked = selectAll;` since it is redundant

holmansze commented on packages/cloud-cognitive/src/components/Datagrid/Datagrid/addons/CustomizeColumns/Columns.js:
```diff
@@ -90,6 +106,27 @@ const Columns = ({
+            <Checkbox
+              wrapperClassName={`${blockClass}__customize-columns-checkbox-wrapper`}
+              className={`${blockClass}__customize-columns-select-all`}
+              checked={getVisibleColumnsCount() == columns.length}
```
In general, when doing this kind of comparison, use `===` instead. `==` performs type conversion, `===` is a strict comparison which takes data type into account.

You may try the following experiment:
- 0 == false (evaluates to true)
- 0 === false (evaluates to false)

kai-liu replied:
Replaced `==` with `===`

holmansze commented on packages/cloud-cognitive/src/components/Datagrid/Datagrid/addons/CustomizeColumns/Columns.js:
```diff
@@ -54,6 +57,19 @@ const Columns = ({
+    let checkboxes = document.querySelectorAll(
+      `.${blockClass}__customize-columns-modal .bx--checkbox`
+    );
+    let selectAll = getVisibleColumnsCount() != columns.length;
```
Use `!==` instead of `!=`.

holmansze commented on packages/cloud-cognitive/src/components/Datagrid/Datagrid/addons/CustomizeColumns/CustomizeColumnsModal.js:
```diff
@@ -62,7 +65,8 @@ const CustomizeColumnsModal = ({
-        return { ...definition, isVisible: value };
+        definition.isVisible = value;
+        return definition;
       }
```
The original code should work and is a better approach, because it creates a new object with the new `isVisible` value (as opposed to mutating the `definition` object).

kai-liu replied:
Reverted to `return { ...definition, isVisible: value };`

kai-liu replied:
Implemented logic that differentiates array parameters and object parameters.

holmansze commented on packages/cloud-cognitive/src/components/Datagrid/Datagrid/addons/CustomizeColumns/CustomizeColumnsModal.js:
```diff
@@ -106,6 +120,8 @@ const CustomizeColumnsModal = ({
       />
       {isOpen && (
         <Columns
+          setVisibleColumns={setVisibleColumns}
```
Do we need to introduce `setVisibleColumns`? Does the existing `setColumnsObject` satisfy our needs?

kai-liu replied:
`setVisibleColumns` is not needed in the context of `Columns`. I have removed it.
